File: src/main/java/maxdistructo/droidbot2/core/Roles.java
package maxdistructo.droidbot2.core;

import maxdistructo.droidbot2.core.message.Message;
import sx.blah.discord.handle.obj.*;

import java.awt.*;
import java.util.EnumSet;
import java.util.List;

public class Roles {
    public static boolean checkForBotAbuse(IMessage message){
        IUser user = message.getAuthor();
        List<IRole> rolesList = user.getRolesForGuild(message.getGuild());
        List<IRole> guildRoles = message.getGuild().getRolesByName("Bot Abuser");
        if(guildRoles.isEmpty()){
            System.out.println("Could not find role Bot Abuser for server " + message.getGuild().getName());
            return false;
        }
        IRole botAbuser = guildRoles.get(0);
        return rolesList.contains(botAbuser);
    }
    public static boolean checkForPayday(IMessage message){
        IUser user = message.getAuthor();
        List<IRole> rolesList = user.getRolesForGuild(message.getGuild());
        List<IRole> guildRoles = message.getGuild().getRolesByName("Payday");
        if(guildRoles.isEmpty()){
            System.out.println("Could not find role Payday for server " + message.getGuild().getName());
            return false;
        }
        IRole payday = guildRoles.get(0);
        return rolesList.contains(payday);
    }
    public static void applyBotAbuser(IMessage message, IUser mentioned){
        List<IRole> guildRoles = message.getGuild().getRolesByName("Bot Abuser");
        if(guildRoles.isEmpty()) {
            System.out.println("Could not find role Bot Abuser for server " + message.getGuild().getName());
        }
        else {
            IRole role = guildRoles.get(0);
            mentioned.addRole(role);
        }
    }
    public static void applyPayday(IMessage message, IUser mentioned){
        IGuild guild = message.getGuild();
        List<IRole> channelList = guild.getRolesByName("Payday");
        IRole role = channelList.get(0);
        mentioned.addRole(role);
    }
    public static void removeBotAbuser(IMessage message, IUser mentioned){
        List<IRole> guildRoles = message.getGuild().getRolesByName("Bot Abuser");
        if(guildRoles.isEmpty()) {
            System.out.println("Could not find role Bot Abuser for server " + message.getGuild().getName());
        }
        else {
        IRole role = guildRoles.get(0);
        mentioned.removeRole(role);
        }
    }
    public static void removePayday(IMessage message, IUser mentioned) {
        List<IRole> guildRoles = message.getGuild().getRolesByName("Payday");
        IRole role = guildRoles.get(0);
        mentioned.removeRole(role);
    }
    public static IRole getRole(IMessage message, String role){
        List<IRole> roles = message.getGuild().getRolesByName(role);
        if(roles.size() != 0){
            return roles.get(0);
        }
        else{
            return null;
        }
    }
    public static void copyRolePerms(IRole copy, IRole paste){
        EnumSet<Permissions> permissions = copy.getPermissions();
        paste.getPermissions().addAll(permissions);
        paste.changePermissions(permissions);
    }
    public static void applyRole(IMessage message, IUser mentioned, String role){
        List<IRole> roleList = message.getGuild().getRolesByName(role);
        if(roleList.isEmpty()){
            Message.sendMessage(message.getChannel(), "The role "+ role + " was not found.");
            Thread.interrupted();
        }
        if(!roleList.isEmpty()) {
            mentioned.addRole(roleList.get(0));
        }
    }
    public static void applyRole(IMessage message, IUser mentioned, long roleL){
        IRole role = message.getGuild().getRoleByID(roleL);
        if(role == null){
            Message.sendMessage(message.getChannel(), "The role "+ roleL + " was not found.");
            Thread.interrupted();
        }
        mentioned.addRole(role);
    }
    public static void changeRolePerm(IMessage message, String perm, String role){
        List<IRole> roleList = message.getGuild().getRolesByName(role);
        if(roleList == null){
            Message.sendMessage(message.getChannel(), "The role "+ role + " was not found.");
            Thread.interrupted();
        }
        IRole roleNew = roleList.get(0);
        EnumSet<Permissions> set = roleNew.getPermissions();
        switch (perm){
            case "administrator":
                set.add(Permissions.ADMINISTRATOR);
            case "muteUsers":
                set.add(Permissions.VOICE_MUTE_MEMBERS);
            case "manageNicknames":
                set.add(Permissions.MANAGE_NICKNAMES);
            case "manageRoles":
                set.add(Permissions.MANAGE_ROLES);
            case "attachFiles":
                set.add(Permissions.ATTACH_FILES);
            case "addReactions":
                set.add(Permissions.ADD_REACTIONS);
            case "banMembers":
                set.add(Permissions.BAN);
            case "changeOwnUsername":
                set.add(Permissions.CHANGE_NICKNAME);
            case "createInvite":
                set.add(Permissions.CREATE_INVITE);
            case "embed":
                set.add(Permissions.EMBED_LINKS);
            case "kick":
                set.add(Permissions.KICK);
            case "manageThisChannel":
                set.add(Permissions.MANAGE_CHANNEL);
            case "manageChannels":
                set.add(Permissions.MANAGE_CHANNELS);
            case "manageEmojis":
                set.add(Permissions.MANAGE_EMOJIS);
            case "manageMessages":
                set.add(Permissions.MANAGE_MESSAGES);
            case "managePermissions":
                set.add(Permissions.MANAGE_PERMISSIONS);
            case "manageServer":
                set.add(Permissions.MANAGE_SERVER);
            case "manageWebhooks":
                set.add(Permissions.MANAGE_WEBHOOKS);
            case "mentionEveryone":
                set.add(Permissions.MENTION_EVERYONE);
            case "readMessageHistory":
                set.add(Permissions.READ_MESSAGE_HISTORY);
            case "readMessages":
                set.add(Permissions.READ_MESSAGES);
            case "sendMessages":
                set.add(Permissions.SEND_MESSAGES);
            case "sendTTSMessages":
                set.add(Permissions.SEND_TTS_MESSAGES);
            case "useExternalEmojis":
                set.add(Permissions.USE_EXTERNAL_EMOJIS);
            case "viewAuditLog":
                set.add(Permissions.VIEW_AUDIT_LOG);
            case "voiceConnect":
                set.add(Permissions.VOICE_CONNECT);
            case "voiceModerator":
                set.add(Permissions.VOICE_DEAFEN_MEMBERS);
                set.add(Permissions.VOICE_MOVE_MEMBERS);
                set.add(Permissions.VOICE_MUTE_MEMBERS);
            case "voiceSpeak":
                set.add(Permissions.VOICE_SPEAK);
            case "voiceUseVad":
                set.add(Permissions.VOICE_USE_VAD);

        }
        roleNew.changePermissions(set);


        }
        public static void changeColor(IRole role, String color){
            Color hex;
            if(color.contains("#")){
                hex = Color.decode(color);
            }
            else{
                hex = Color.decode("#" + color);
            }
            role.changeColor(hex);
        }
        public static void makeNewRole(IGuild guild, String roleName, boolean hoist, boolean mentionable){
            RoleBuilder rb = new RoleBuilder(guild);
            rb.withName(roleName);
            rb.setHoist(hoist);
            rb.setMentionable(mentionable);
            rb.build();
        }
    }

File: src/main/java/maxdistructo/droidbot2/core/Roles.kt
package maxdistructo.droidbot2.core

import maxdistructo.droidbot2.core.message.Message
import sx.blah.discord.handle.obj.*
import sx.blah.discord.util.RoleBuilder

import java.awt.*
import java.util.EnumSet

object Roles {
    fun checkForBotAbuse(message: IMessage): Boolean {
        val user = message.author
        val rolesList = user.getRolesForGuild(message.guild)
        val guildRoles = message.guild.getRolesByName("Bot Abuser")
        if (guildRoles.isEmpty()) {
            println("Could not find role Bot Abuser for server " + message.guild.name)
            return false
        }
        val botAbuser = guildRoles[0]
        return rolesList.contains(botAbuser)
    }

    fun checkForPayday(message: IMessage): Boolean {
        val user = message.author
        val rolesList = user.getRolesForGuild(message.guild)
        val guildRoles = message.guild.getRolesByName("Payday")
        if (guildRoles.isEmpty()) {
            println("Could not find role Payday for server " + message.guild.name)
            return false
        }
        val payday = guildRoles[0]
        return rolesList.contains(payday)
    }

    fun applyBotAbuser(message: IMessage, mentioned: IUser) {
        val guildRoles = message.guild.getRolesByName("Bot Abuser")
        if (guildRoles.isEmpty()) {
            println("Could not find role Bot Abuser for server " + message.guild.name)
        } else {
            val role = guildRoles[0]
            mentioned.addRole(role)
        }
    }

    fun applyPayday(message: IMessage, mentioned: IUser) {
        val guild = message.guild
        val channelList = guild.getRolesByName("Payday")
        val role = channelList[0]
        mentioned.addRole(role)
    }

    fun removeBotAbuser(message: IMessage, mentioned: IUser) {
        val guildRoles = message.guild.getRolesByName("Bot Abuser")
        if (guildRoles.isEmpty()) {
            println("Could not find role Bot Abuser for server " + message.guild.name)
        } else {
            val role = guildRoles[0]
            mentioned.removeRole(role)
        }
    }

    fun removePayday(message: IMessage, mentioned: IUser) {
        val guildRoles = message.guild.getRolesByName("Payday")
        val role = guildRoles[0]
        mentioned.removeRole(role)
    }

    fun getRole(message: IMessage, role: String): IRole? {
        val roles = message.guild.getRolesByName(role)
        return if (roles.size != 0) {
            roles[0]
        } else {
            null
        }
    }

    fun copyRolePerms(copy: IRole, paste: IRole) {
        val permissions = copy.permissions
        paste.permissions.addAll(permissions)
        paste.changePermissions(permissions)
    }

    fun applyRole(message: IMessage, mentioned: IUser, role: String) {
        val roleList = message.guild.getRolesByName(role)
        if (roleList.isEmpty()) {
            Message.sendMessage(message.channel, "The role $role was not found.")
            Thread.interrupted()
        }
        if (!roleList.isEmpty()) {
            mentioned.addRole(roleList[0])
        }
    }

    fun applyRole(message: IMessage, mentioned: IUser, roleL: Long) {
        val role = message.guild.getRoleByID(roleL)
        if (role == null) {
            Message.sendMessage(message.channel, "The role $roleL was not found.")
            Thread.interrupted()
        }
        mentioned.addRole(role)
    }

    fun changeRolePerm(message: IMessage, perm: String, role: String) {
        val roleList = message.guild.getRolesByName(role)
        if (roleList == null) {
            Message.sendMessage(message.channel, "The role $role was not found.")
            Thread.interrupted()
        }
        val roleNew = roleList!![0]
        val set = roleNew.permissions
        when (perm) {
            "administrator" -> {
                set.add(Permissions.ADMINISTRATOR)
            }
            "muteUsers" -> {
                set.add(Permissions.VOICE_MUTE_MEMBERS)
            }
            "manageNicknames" -> {
                set.add(Permissions.MANAGE_NICKNAMES)
            }
            "manageRoles" -> {
                set.add(Permissions.MANAGE_ROLES)
            }
            "attachFiles" -> {
                set.add(Permissions.ATTACH_FILES)
            }
            "addReactions" -> {
                set.add(Permissions.ADD_REACTIONS)
            }
            "banMembers" -> {
                set.add(Permissions.BAN)
            }
            "changeOwnUsername" -> {
                set.add(Permissions.CHANGE_NICKNAME)
            }
            "createInvite" -> {
                set.add(Permissions.CREATE_INVITE)
            }
            "embed" -> {
                set.add(Permissions.EMBED_LINKS)
            }
            "kick" -> {
                set.add(Permissions.KICK)
            }
            "manageThisChannel" -> {
                set.add(Permissions.MANAGE_CHANNEL)
            }
            "manageChannels" -> {
                set.add(Permissions.MANAGE_CHANNELS)
            }
            "manageEmojis" -> {
                set.add(Permissions.MANAGE_EMOJIS)
            }
            "manageMessages" -> {
                set.add(Permissions.MANAGE_MESSAGES)
            }
            "managePermissions" -> {
                set.add(Permissions.MANAGE_PERMISSIONS)
            }
            "manageServer" -> {
                set.add(Permissions.MANAGE_SERVER)
            }
            "manageWebhooks" -> {
                set.add(Permissions.MANAGE_WEBHOOKS)
            }
            "mentionEveryone" -> {
                set.add(Permissions.MENTION_EVERYONE)
            }
            "readMessageHistory" -> {
                set.add(Permissions.READ_MESSAGE_HISTORY)
            }
            "readMessages" -> {
                set.add(Permissions.READ_MESSAGES)
            }
            "sendMessages" -> {
                set.add(Permissions.SEND_MESSAGES)
            }
            "sendTTSMessages" -> {
                set.add(Permissions.SEND_TTS_MESSAGES)
            }
            "useExternalEmojis" -> {
                set.add(Permissions.USE_EXTERNAL_EMOJIS)
            }
            "viewAuditLog" -> {
                set.add(Permissions.VIEW_AUDIT_LOG)
            }
            "voiceConnect" -> {
                set.add(Permissions.VOICE_CONNECT)
            }
            "voiceModerator" -> {
                set.add(Permissions.VOICE_DEAFEN_MEMBERS)
            }
            "voiceSpeak" -> {
                set.add(Permissions.VOICE_SPEAK)
            }
            "voiceUseVad" -> set.add(Permissions.VOICE_USE_VAD)
        }
        roleNew.changePermissions(set)


    }

    fun changeColor(role: IRole, color: String) {
        val hex: Color
        if (color.contains("#")) {
            hex = Color.decode(color)
        } else {
            hex = Color.decode("#$color")
        }
        role.changeColor(hex)
    }

    fun makeNewRole(guild: IGuild, roleName: String, hoist: Boolean, mentionable: Boolean) {
        val rb = RoleBuilder(guild)
        rb.withName(roleName)
        rb.setHoist(hoist)
        rb.setMentionable(mentionable)
        rb.build()
    }
}

